ddl/owner.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	goctx "golang.org/x/net/context"
+)
+
+// ChangeOwnerInNewWay is used for testing.
+var ChangeOwnerInNewWay = false
+
+const (
+	ddlOwnerKey               = "/tidb/ddl/owner"
+	bgOwnerKey                = "/tidb/ddl/bg/owner"
+	newSessionDefaultRetryCnt = 3
+)
+
+type worker struct {
+	ddlOwner    int32
+	bgOwner     int32
+	ddlID       string
+	etcdClient  *clientv3.Client
+	etcdSession *concurrency.Session
+}
+
+func (w *worker) isOwner() bool {
+	return atomic.LoadInt32(&w.ddlOwner) == 1
+}
+
+func (w *worker) setOwner(isOwner bool) {
+	if isOwner {
+		atomic.StoreInt32(&w.ddlOwner, 1)
+	} else {
+		atomic.StoreInt32(&w.ddlOwner, 0)
+	}
+}
+
+func (w *worker) isBgOwner() bool {
+	return atomic.LoadInt32(&w.bgOwner) == 1
+}
+
+func (w *worker) setBgOwner(isOwner bool) {
+	if isOwner {
+		atomic.StoreInt32(&w.bgOwner, 1)
+	} else {
+		atomic.StoreInt32(&w.bgOwner, 0)
+	}
+}
+
+func (w *worker) newSession(ctx goctx.Context, retryCnt int) error {
+	var err error
+	for i := 0; i < retryCnt; i++ {
+		w.etcdSession, err = concurrency.NewSession(w.etcdClient, concurrency.WithContext(ctx))
+		if err != nil {
+			log.Warnf("[ddl] failed to new session, err %v", err)
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+		break
+	}
+	return errors.Trace(err)
+}
+
+func (d *ddl) campaignOwners(ctx goctx.Context) error {
+	err := d.worker.newSession(ctx, newSessionDefaultRetryCnt)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	d.wait.Add(2)
+	ddlCtx, _ := goctx.WithCancel(ctx)
+	go d.campaignLoop(ddlCtx, ddlOwnerKey)
+
+	bgCtx, _ := goctx.WithCancel(ctx)
+	go d.campaignLoop(bgCtx, bgOwnerKey)
+	return nil
+}
+
+func (d *ddl) campaignLoop(ctx goctx.Context, key string) {
+	defer d.wait.Done()
+	worker := d.worker
+	for {
+		if d.isClosed() {
+			return
+		}
+
+		elec := concurrency.NewElection(worker.etcdSession, key)
+		err := elec.Campaign(ctx, worker.ddlID)
+		if err != nil {
+			log.Infof("[ddl] worker %s failed to campaign, err %v", worker.ddlID, err)
+			continue
+		}
+
+		// Get owner information.
+		resp, err := elec.Leader(ctx)
+		if err != nil {
+			// If no leader elected currently, it returns ErrElectionNoLeader.
+			log.Infof("[ddl] failed to get leader, err %v", err)
+			continue
+		}
+		leader := string(resp.Kvs[0].Value)
+		log.Info("[ddl] %s worker is %s, owner is %v", key, worker.ddlID, leader)
+		if leader == worker.ddlID {
+			worker.setOwnerVal(key, true)
+		} else {
+			log.Warnf("[ddl] worker %s isn't the owner", worker.ddlID)
+			continue
+		}
+
+		// TODO: Use content instead of quitCh.
+		worker.watchOwner(ctx, string(resp.Kvs[0].Key))
+		worker.setOwnerVal(key, false)
+		d.hookMu.Lock()
+		d.hook.OnWatched(ctx)
+		d.hookMu.Unlock()
+
+		select {
+		case <-worker.etcdSession.Done():
+			// TODO: Create session again?
+			log.Warnf("etcd session is done.")
```

Comment:
and we don't need to check this here. We know the session has been done in `watchOwner`. So we can retry forever at there.